trieste/ask_tell_optimization.py

```diff
@@ -0,0 +1,290 @@
+        model_specs = cast(Dict[str, ModelSpec], model_specs)
+
+        if datasets.keys() != model_specs.keys():
+            raise ValueError(
+                f"datasets and model_specs should contain the same keys. Got {datasets.keys()} and"
+                f" {model_specs.keys()} respectively."
+            )
+
+        self._datasets = datasets
+        self._models = map_values(create_model, model_specs)
+
+        if acquisition_rule is None:
+            if self._datasets.keys() != {OBJECTIVE}:
+                raise ValueError(
+                    f"Default acquisition rule EfficientGlobalOptimization requires tag"
+                    f" {OBJECTIVE!r}, got keys {self._datasets.keys()}"
+                )
+
+            self._acquisition_rule = cast(
+                AcquisitionRule[TensorType, SP], EfficientGlobalOptimization()
+            )
+        else:
+            self._acquisition_rule = acquisition_rule
+
+        if fit_model:
+            for tag, model in self._models.items():
+                dataset = datasets[tag]
+                model.update(dataset)
+                model.optimize(dataset)
+
+    def __repr__(self) -> str:
+        """Print-friendly string representation"""
+        return f"""AskTellOptimizer({self._search_space!r}, {self._datasets!r},
+               {self._models!r}, {self._acquisition_rule!r}), "
+               {self._acquisition_state!r}"""
+
+    @classmethod
+    def from_record(
+        cls,
+        record: Record[S],
+        search_space: SP,
+        acquisition_rule: AcquisitionRule[TensorType | State[S | None, TensorType], SP]
+        | None = None,
+    ) -> AskTellOptimizer[SP]:
+        """Creates new :class:`~AskTellOptimizer` instance from provided optimization state.
+
+        :param record: Optimization state record.
+        :param search_space: The space over which to search for the next query point.
+        :param acquisition_rule: The acquisition rule, which defines how to search for a new point
+            on each optimization step. Defaults to
+            :class:`~trieste.acquisition.rule.EfficientGlobalOptimization` with default
+            arguments.
+        """
+        # we are recovering previously saved optimization state
+        # so the model was already trained
+        # thus there is no need to train it again
+        return cls(
+            search_space,
+            record.datasets,
+            record.models,
+            acquisition_rule=acquisition_rule,
+            acquisition_state=record.acquisition_state,
+            fit_model=False,
+        )
+
+    def to_record(self) -> Record[S]:
+        """Collects the current state of the optimization, which includes datasets,
+        models and acquisition state (if applicable).
+
+        :return: An optimization state record.
+        """
+        models_copy = copy.deepcopy(self._models)
+        acquisition_state_copy = copy.deepcopy(self._acquisition_state)
```

Comment:
is this meaningful only when acquisition function is stateful? even if this passes for stateless acquisition functions, would it mean that you couldn't continue optimization from a `Record`?
seems like it might be worth having a warning about this in the docstring